core/io.go

```diff
@@ -0,0 +1,98 @@
+package core
+
+import (
+	"bytes"
+	"errors"
+	"io"
+
+	"github.com/dicedb/dice/config"
+)
+
+type RESPParser struct {
+	c   io.ReadWriter
+	buf *bytes.Buffer
+}
+
+func NewRESPParser(c io.ReadWriter) *RESPParser {
+	return NewRESPParserWithBytes(c, []byte{})
+}
+
+func NewRESPParserWithBytes(c io.ReadWriter, initBytes []byte) *RESPParser {
+	var b []byte
+	var buf *bytes.Buffer = bytes.NewBuffer(b)
+	buf.Write(initBytes)
+	return &RESPParser{
+		c:   c,
+		buf: buf,
+	}
+}
+
+func (rp *RESPParser) DecodeOne() (interface{}, error) {
+	// assigning temporary buffer to read 512 bytes in one shot
+	// and reading them in a loop until we have all the data
+	// we want.
+	// note: the size 512 is arbitrarily chosen, and we can put
+	// a decent thought into deciding the optimal value (in case it affects the perf)
+	var buf []byte = make([]byte, config.IOBufferLength)
+
+	// keep reading the bytes from the buffer until the first /r is found
+	// note: there may be extra bytes read over the socket post /r
+	// but that is fine.
+	for {
+		n, err := rp.c.Read(buf)
+		if n <= 0 {
+			break
+		}
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		rp.buf.Write(buf[:n])
+
+		if bytes.Contains(buf, []byte{'\r', '\n'}) {
```

Comment:
Thanks a ton, @asutosh97 for bringing up this brilliant point. Let me add my views on this

When a client wants to leverage command pipelining, it will be looking to aggressively send data to the server, and hence by continuously reading the socket, we will be buffering everything that is available through the `DecodeMultiple` function that is invoked during `readCommands`.

Because reading commands happens prior to the evaluation, we are anyway buffering the entire input before triggering the evaluation. During the evaluation, we will be iterating over the commands and evaluating them one by one.

I may be wrong, but I do not see this as becoming a problem. But I propose we benchmark with a very long command (once we reach a decent shape) and see. what say?